### ISSUE TITLE
Unnecessary import error for used imports with * at the end

### ISSUE THREAD

**Issue by illuzor** [pinterest/ktlint#1277]
## Expected Behavior
No error

## Observed Behavior
Error

## Steps to Reproduce
Disable **no-wildcard-imports** rule and create a file with wildcard import

## Your Environment

* Version of ktlint used: 0.43.0
* Name and version (or code for custom task) of integration used (Gradle plugin, Maven plugin, command line, custom Gradle task): gradlew:ktlintCheck
* Version of Gradle used (if applicable): 7.2
* Operating System and version: Windows 10
* Link to your project (if it's a public repository): ...

.editorconfig:
```
[*.{kt,kts}]
disabled_rules=no-wildcard-imports
```
Code:
``` kotlin
import test.*

fun main() {
    Test1()
    Test2()
    Test3()
}
```

Output:
```
> Task :ktlintMainSourceSetCheck FAILED
D:\_sources\IDEA\tests\ktlint_error\src\main\kotlin\demo.kt:1:1 Unnecessary import
3 actionable tasks: 1 executed, 2 up-to-date

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':ktlintMainSourceSetCheck'.
> A failure occurred while executing org.jlleitschuh.gradle.ktlint.worker.ConsoleReportWorkAction
   > KtLint found code style violations. Please see the following reports:
     - D:\_sources\IDEA\tests\ktlint_error\build\reports\ktlint\ktlintMainSourceSetCheck\ktlintMainSourceSetCheck.txt

```
Fails only if run by the [plugin](https://github.com/JLLeitschuh/ktlint-gradle/issues) task **gradlew ktlintCheck**.
Does not fail if run ktlint from command line

Sample project: [ktlint_error.zip](https://github.com/pinterest/ktlint/files/7522360/ktlint_error.zip)

**Comment by romtsn** [comment]
you're disabling the wrong rule, it should actually be `no-unused-imports`

**Comment by illuzor** [comment]
@romtsn no, I`m disabling correct rule.
no-unused-imports should fail when wildcard import is unused, but in my case it is used and should not fail

**Comment by romtsn** [comment]
Ah, alright, then it's indeed a bug

**Comment by paul-dingemans** [comment]
@illuzor Please copy paste the code samples and output in the issue instead of an image. Also you should provide the expected code.

**Comment by paul-dingemans** [comment]
With file `Foo.kt` and content:
```
import some.path.*

interface Foo : Bar
```

and `.editorconfig`:
```
[*.{kt,kts}]
disabled_rules=no-wildcard-imports
```

ktlint does not fail for me as expected. However when I remove the disabled_rules from the `.editorconfig` it does fail.

**Comment by illuzor** [comment]
@paul-dingemans done

**Comment by paul-dingemans** [comment]
@romtsn This problem has been introduced by fixing #1256. I am stuck with trying to resolve this issue without reverting #1256. 

The basic problem is that due to #1256 the import in code below is removed as no reference is found to package `test`.
```
import test.*

fun main() {
    Test() // defined in package test
}
```
So when visiting the CALL_EXPRESSION `Test()` or one level deeper at the IDENTIFIER node `Test` , I need to extract information from the psi node about the import that is required or the fully qualified name of the class. But in those nodes, I can not find any helpful information. Do you have any idea how to approach this?

**Comment by GabrielLasso** [comment]
@romtsn I had the same problem, then I created an example to reproduce it and I was going to report it, but I found this issue reporting the same problem.

Here is the example I created, I hope it is useful somehow

https://github.com/GabrielLasso/ktlint-bug-example

**Comment by henrik242** [comment]
@paul-dingemans Any progress on this? This bug removes in-use star imports when I use the formatter [through kotlinter](https://github.com/jeremymailen/kotlinter-gradle/issues/227).

**Comment by paul-dingemans** [comment]
Sorry no progress. Waiting for response of @romtsn

**Comment by romtsn** [comment]
I'd need to look into that, but from the top of my head we probably can't extract this info easily. the proper way would be to depend on the compiler analysis, but that requires quite some setup

**Comment by henrik242** [comment]
My suggestion would be to revert the change that introduced this regression, and work from there.  This way projects like [kotlinter](https://github.com/jeremymailen/kotlinter-gradle/issues/227) can continue to track the latest ktlint.

**Comment by Mehly** [comment]
Any updates on this issue yet?

We're unable to upgrade to the latest version of ktlint due to this issue because Android Studio by default enforces wildcard-imports after a certain threshold of imports from the same package has been reached.

**Comment by jakoss** [comment]
@Mehly You can turn that off in IntelliJ settings:
![image](https://user-images.githubusercontent.com/3933348/154247138-06d6f9f0-1ad3-4237-b25d-bc4e0d22f4ec.png)

But I do agree, we're still on 0.41.0 since this is the last version that does work for us just fine

**Comment by romtsn** [comment]
@Mehly you can also disable it through .editorconfig `ij_kotlin_name_count_to_use_star_import = 2147483647`

**Comment by paul-dingemans** [comment]
@InTheCloudDan can you please try to minimize your example and just paste it here? Also, it would be better to use an EOL comment to prevent removal of the import on that specific line. Now you have disabled it for the remainder of the file.

**Comment by paul-dingemans** [comment]
> @romtsn This problem has been introduced by fixing #1256. I am stuck with trying to resolve this issue without reverting #1256.
> 
> The basic problem is that due to #1256 the import in code below is removed as no reference is found to package `test`.
> 
> ```
> import test.*
> 
> fun main() {
>     Test() // defined in package test
> }
> ```
> 
> So when visiting the CALL_EXPRESSION `Test()` or one level deeper at the IDENTIFIER node `Test` , I need to extract information from the psi node about the import that is required or the fully qualified name of the class. But in those nodes, I can not find any helpful information. Do you have any idea how to approach this?

Solution of #1256 has been reverted as it leads to compilation failures when required wildcard important are removed. Downside is that some unused wilcard imports may not be detected. But that is not considered as harmfull.
